File: rover_explorer_ros2/rover_explorer_ros2/obstacle_grid_node.py
from __future__ import annotations

import math

import rclpy
from nav_msgs.msg import OccupancyGrid
from rclpy.node import Node
from sensor_msgs.msg import Range

from rover_explorer.localize import RoverPose as CorePose
from rover_explorer.obstacle import ObstacleGrid
from rover_explorer_ros2.msg import RoverPose

from .common import declare_transform_parameters, transform_from_node


class ObstacleGridNode(Node):
    def __init__(self) -> None:
        super().__init__("obstacle_grid_node")
        declare_transform_parameters(self)
        self.declare_parameter("camera_width", 640)
        self.declare_parameter("camera_height", 480)
        self.declare_parameter("map_cols", 12)
        self.declare_parameter("map_rows", 8)
        self.declare_parameter("obstacle_ttl_cycles", 40)
        self.declare_parameter("cm_per_translation_pulse", 10.0)
        self.declare_parameter("maximum_mapping_distance_cm", 150)
        self.declare_parameter("map_frame_id", "map")
        self._grid = ObstacleGrid(
            (int(self.get_parameter("camera_height").value), int(self.get_parameter("camera_width").value), 3),
            int(self.get_parameter("map_cols").value),
            int(self.get_parameter("map_rows").value),
            int(self.get_parameter("obstacle_ttl_cycles").value),
        )
        self._pose: CorePose | None = None
        self._cycle = 0
        self._publisher = self.create_publisher(OccupancyGrid, "/rover/occupancy_grid", 10)
        self.create_subscription(RoverPose, "/rover/pose", self._on_pose, 10)
        self.create_subscription(Range, "/rover/sonar", self._on_sonar, 10)

    def _on_pose(self, message: RoverPose) -> None:
        self._pose = CorePose(
            (message.centre.x, message.centre.y),
            message.heading if message.has_heading else None,
            message.confidence,
        )

    def _on_range(self, message: Range, offset: float) -> None:
        if self._pose is None or not math.isfinite(message.range):
            return
        self._cycle += 1
        transform = transform_from_node(self)
        heading = self._pose.heading
        if heading is None:
            dx, dy = transform.direction(self._pose)
            heading = math.atan2(dy, dx)
        distance_cm = message.range * 100.0
        maximum = float(self.get_parameter("maximum_mapping_distance_cm").value)
        distance_px = min(distance_cm, maximum) * transform.px_per_forward_pulse / float(
            self.get_parameter("cm_per_translation_pulse").value
        )
        self._grid.observe_ray(
            self._pose.centre, heading + offset, distance_px,
            hit=distance_cm < maximum, cycle=self._cycle,
        )
        self._publish()

    def _on_sonar(self, message: Range) -> None:
        offsets = {
            "sonar_front": 0.0,
            "sonar_left": math.radians(50),
            "sonar_right": -math.radians(50),
        }
        if message.header.frame_id in offsets:
            self._on_range(message, offsets[message.header.frame_id])

    def _publish(self) -> None:
        message = OccupancyGrid()
        message.header.stamp = self.get_clock().now().to_msg()
        message.header.frame_id = str(self.get_parameter("map_frame_id").value)
        message.info.width = self._grid.cols
        message.info.height = self._grid.rows
        message.info.resolution = 1.0
        message.info.origin.orientation.w = 1.0
        occupied = self._grid.occupied()
        message.data = [
            100 if (col, row) in occupied else 0
            for row in range(self._grid.rows)
            for col in range(self._grid.cols)
        ]
        self._publisher.publish(message)


def main(args=None) -> None:
    rclpy.init(args=args)
    node = ObstacleGridNode()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: rover_explorer_ros2/src/obstacle_grid_node.cpp
#include <algorithm>
#include <cmath>
#include <cstdint>
#include <memory>
#include <optional>
#include <string>

#include "nav_msgs/msg/occupancy_grid.hpp"
#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/range.hpp"

#include "rover_explorer_ros2/deterministic_logic.hpp"
#include "rover_explorer_ros2/msg/rover_pose.hpp"

namespace rover_explorer_ros2
{

class ObstacleGridNode : public rclcpp::Node
{
public:
  ObstacleGridNode()
  : Node("obstacle_grid_node")
  {
    declare_parameter("px_per_forward_pulse", 35.0);
    declare_parameter("radians_per_turn_pulse", 0.35);
    declare_parameter("forward_axis_x", 1.0);
    declare_parameter("forward_axis_y", 0.0);
    declare_parameter("camera_width", 640);
    declare_parameter("camera_height", 480);
    declare_parameter("map_cols", 12);
    declare_parameter("map_rows", 8);
    declare_parameter("obstacle_ttl_cycles", 40);
    declare_parameter("cm_per_translation_pulse", 10.0);
    declare_parameter("maximum_mapping_distance_cm", 150);
    declare_parameter("map_frame_id", "map");
    grid_ = std::make_unique<ObstacleGrid>(
      static_cast<int>(get_parameter("camera_width").as_int()),
      static_cast<int>(get_parameter("camera_height").as_int()),
      static_cast<int>(get_parameter("map_cols").as_int()),
      static_cast<int>(get_parameter("map_rows").as_int()),
      static_cast<int>(get_parameter("obstacle_ttl_cycles").as_int()));

    const auto qos = rclcpp::QoS(rclcpp::KeepLast(10));
    publisher_ = create_publisher<nav_msgs::msg::OccupancyGrid>("/rover/occupancy_grid", qos);
    pose_subscription_ = create_subscription<msg::RoverPose>(
      "/rover/pose", qos,
      [this](const msg::RoverPose::SharedPtr message) {on_pose(*message);});
    sonar_subscription_ = create_subscription<sensor_msgs::msg::Range>(
      "/rover/sonar", qos,
      [this](const sensor_msgs::msg::Range::SharedPtr message) {on_sonar(*message);});
  }

private:
  static constexpr double kPi = 3.14159265358979323846;

  void on_pose(const msg::RoverPose & message)
  {
    pose_ = Pose2D{
      message.centre.x,
      message.centre.y,
      message.has_heading ? std::optional<double>(message.heading) : std::nullopt,
      message.confidence};
  }

  BodyToImage transform() const
  {
    return BodyToImage{
      get_parameter("px_per_forward_pulse").as_double(),
      get_parameter("radians_per_turn_pulse").as_double(),
      get_parameter("forward_axis_x").as_double(),
      get_parameter("forward_axis_y").as_double()};
  }

  void on_sonar(const sensor_msgs::msg::Range & message)
  {
    if (!pose_.has_value() || !std::isfinite(message.range)) {
      return;
    }
    double offset = 0.0;
    if (message.header.frame_id == "sonar_left") {
      offset = 50.0 * kPi / 180.0;
    } else if (message.header.frame_id == "sonar_right") {
      offset = -50.0 * kPi / 180.0;
    } else if (message.header.frame_id != "sonar_front") {
      return;
    }

    ++cycle_;
    const auto body_transform = transform();
    auto heading = pose_->heading;
    if (!heading.has_value()) {
      const auto [dx, dy] = body_transform.direction(*pose_);
      heading = std::atan2(dy, dx);
    }
    const auto distance_cm = static_cast<double>(message.range) * 100.0;
    const auto maximum = static_cast<double>(
      get_parameter("maximum_mapping_distance_cm").as_int());
    const auto cm_per_pulse = get_parameter("cm_per_translation_pulse").as_double();
    const auto distance_px = std::min(distance_cm, maximum) *
      body_transform.px_per_forward_pulse / cm_per_pulse;
    grid_->observe_ray(
      pose_->x, pose_->y, *heading + offset, distance_px,
      distance_cm < maximum, cycle_);
    publish_grid();
  }

  void publish_grid()
  {
    nav_msgs::msg::OccupancyGrid output;
    output.header.stamp = get_clock()->now();
    output.header.frame_id = get_parameter("map_frame_id").as_string();
    output.info.width = static_cast<std::uint32_t>(grid_->cols());
    output.info.height = static_cast<std::uint32_t>(grid_->rows());
    output.info.resolution = 1.0F;
    output.info.origin.orientation.w = 1.0;
    const auto occupied = grid_->occupied();
    output.data.reserve(static_cast<std::size_t>(grid_->cols() * grid_->rows()));
    for (int row = 0; row < grid_->rows(); ++row) {
      for (int col = 0; col < grid_->cols(); ++col) {
        output.data.push_back(
          occupied.count({col, row}) > 0 ? static_cast<std::int8_t>(100) : 0);
      }
    }
    publisher_->publish(output);
  }

  std::unique_ptr<ObstacleGrid> grid_;
  std::optional<Pose2D> pose_;
  std::int64_t cycle_{0};
  rclcpp::Publisher<nav_msgs::msg::OccupancyGrid>::SharedPtr publisher_;
  rclcpp::Subscription<msg::RoverPose>::SharedPtr pose_subscription_;
  rclcpp::Subscription<sensor_msgs::msg::Range>::SharedPtr sonar_subscription_;
};

}  // namespace rover_explorer_ros2

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<rover_explorer_ros2::ObstacleGridNode>());
  rclcpp::shutdown();
  return 0;
}
